libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialogBot.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs.Adaptive.Generators;
+using Microsoft.Bot.Builder.Dialogs.Declarative.Resources;
+using Microsoft.Bot.Builder.Skills;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// An <see cref="IBot"/> implementation that hosts an <see cref="AdaptiveDialog"/>.
+    /// </summary>
+    public class AdaptiveDialogBot : IBot
+    {
+        private const string DefaultLocaleKey = "defaultLocale";
+        private const string RootDialogKey = "defaultRootDialog";
+
+        private const string DefaultLocale = "en-US";
+        private const string DefaultLg = "main.lg";
+
+        private static ConcurrentDictionary<ResourceExplorer, LanguageGeneratorManager> _languageGeneratorManagers = new ConcurrentDictionary<ResourceExplorer, LanguageGeneratorManager>();
+
+        private readonly string _defaultLocale;
+        private readonly string _rootDialogId;
+
+        private readonly ILogger _logger;
+        private readonly ResourceExplorer _resourceExplorer;
+        private readonly IStorage _storage;
+        private readonly BotFrameworkClient _botFrameworkClient;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveDialogBot"/> class.
+        /// </summary>
+        /// <param name="configuration">An <see cref="IConfiguration"/> instance.</param>
+        /// <param name="logger">An <see cref="ILogger"/> instance.</param>
+        /// <param name="storage">The <see cref="IStorage"/> implementation to use for this <see cref="Dialog"/>.</param>
+        /// <param name="resourceExplorer">The Bot Builder <see cref="ResourceExplorer"/> to load the <see cref="Dialog"/> from.</param>
+        /// <param name="botFrameworkClient">A <see cref="BotFrameworkClient"/> for making calls to Bot Builder Skills.</param>
+        public AdaptiveDialogBot(
```

Comment:
I think it may be good to add explicit unit tests for the constructor assumptions:

![image](https://user-images.githubusercontent.com/12264946/110896205-1eb2be00-82c9-11eb-8102-cc1405b41dad.png)

Reply:
So the unit test is reasonably good as far as it goes - I was slightly surprised how much it covered!
Note the constructor is also inside the "ACT" part of the test rather than the ARRANGE, that's significant.

I can add other constructor specific tests, but I wanted to see what you thought about this single constructor with its inherent coupling to defaulting and configuration.